modules/permutiveIdSystem.js

```diff
@@ -0,0 +1,153 @@
+  const id = {}
+  if (data && typeof data === 'object') {
+    const now = Date.now()
+    for (const [idName, value] of Object.entries(data)) {
+      if (PRIMARY_IDS.includes(idName) && value.userId) {
+        if (!value.expiryTime || value.expiryTime > now) {
+          id[idName] = value.userId
+        }
+      }
+    }
+  }
+  return id
+}
+
+/**
+ * Catch and log errors
+ * @param {function} fn - Function to safely evaluate
+ */
+function makeSafe (fn) {
+  try {
+    return fn()
+  } catch (e) {
+    logger.logError(e)
+  }
+}
+
+const waitAndRetrieveFromSdk = (timeoutMs) =>
+  new Promise(
+    resolve => {
+      const fallback = setTimeout(() => {
+        logger.logInfo('timeout expired waiting for SDK - attempting read from local storage again')
+        resolve(readFromSdkLocalStorage())
+      }, timeoutMs)
+      if (window.permutive) {
+        window.permutive.ready(() => makeSafe(() => {
+          logger.logInfo('Permutive SDK is ready')
+          const sdkIdentityManagerApi = window.permutive.addon('identityManager')
+          if (sdkIdentityManagerApi && 'prebid' in sdkIdentityManagerApi && 'onReady' in sdkIdentityManagerApi.prebid) {
+            sdkIdentityManagerApi.prebid.onReady((ids) => {
+              logger.logInfo('Permutive SDK has provided ids')
+              resolve(ids)
+              clearTimeout(fallback)
+            })
+          } else {
+            logger.logError('Permutive SDK initialised but identity manager prebid api not present')
+          }
+        }))
+      }
+    }
+  )
+
+/** @type {Submodule} */
+export const permutiveIdSubmodule = {
+  /**
+   * used to link submodule with config
+   * @type {string}
+   */
+  name: MODULE_NAME,
+
+  /**
+   * decode the stored id value for passing to bid requests
+   * @function decode
+   * @param {(Object|string)} value
+   * @param {SubmoduleConfig|undefined} config
+   * @returns {(Object|undefined)}
+   */
+  decode(value, config) {
+    return value
+  },
+
+  /**
+   * performs action to obtain id and return a value in the callback's response argument
+   * @function getId
+   * @param {SubmoduleConfig} submoduleConfig
+   * @param {ConsentData} consentData
+   * @param {(Object|undefined)} cacheIdObj
+   * @returns {IdResponse|undefined}
+   */
+  getId(submoduleConfig, consentData, cacheIdObj) {
+    const id = readFromSdkLocalStorage()
+    if (Object.entries(id).length > 0) {
+      logger.logInfo('found id in sdk storage')
+      return { id }
+    } else if ('params' in submoduleConfig && submoduleConfig.params.ajaxTimeout) {
+      logger.logInfo('failed to find id in sdk storage - waiting for sdk')
+      // Is ajaxTimeout an appropriate timeout to use here?
+      return { callback: (done) => waitAndRetrieveFromSdk(submoduleConfig.params.ajaxTimeout).then(done) }
+    } else {
+      logger.logInfo('failed to find id in sdk storage and no wait time specified')
+    }
+  },
+
+  primaryIds: PRIMARY_IDS,
+
+  eids: {
```

Comment:
Question to prebid maintainers, if we were to make this a getter property such that we could read the `eids` config directly from our SDK and use what we have defined here as a fallback, would this propagate to other parts of the identity system as it would be called frequently? Or is the initial object returned stored by reference somewhere?

i.e.
```javascript
get eids() {
  const data = read.from.permutive.sdk
  if (data) {
    return data
  }

  return { id5id: ... }
}
```